### PR TITLE
[DTMF Dialer] Show hide the dtmf dialer when calling someone

### PR DESCRIPTION
**Pull request by dmceachernmsft**
# What

- updates Dialpad API to have more distinct modes as per designs
- Shows and hides the dialpad when people join the call
# Why

Allows for seamless behavior for the composite to show the dialpad and dismiss it when the user has an agent in the call.
- will not clear the dialpad after this initial dismiss to signify that the user has joined the call

https://skype.visualstudio.com/SPOOL/_workitems/edit/3580335
# How Tested

Locally see video

https://github.com/Azure/communication-ui-library/assets/94866715/9060f480-7dc9-4f24-9cba-b28bf2cb5825

### CODE REVIEW

File: packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { ActiveErrorMessage, ErrorBar, ParticipantMenuItemsCallback } from '@int
 import { VideoGalleryLayout } from '@internal/react-components';
 import React from 'react';
 /* @conditional-compile-remove(dtmf-dialer) */
-import { useState } from 'react';
+import { useState, useEffect } from 'react';
```
Comment (edwardlee-msft) [inline packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx:12]: this useEffect may cause an error if it's only being used in a conditionally compiled way in the rest of the file.

File: packages/react-composites/src/composites/CallComposite/utils/MediaGalleryUtils.ts [hunk 2]
```diff
@@ -160,3 +166,21 @@ export const createAnnouncementString = (
...
+ * @returns whether the dialer should be the gallery content or not
+ */
+export const showDtmfDialer = (callees?: CommunicationIdentifier[]): boolean => {
+  return !!callees?.filter((callee) => {
+    if (isPhoneNumberIdentifier(callee)) {
+      return true;
+    } else if (isMicrosoftTeamsAppIdentifier(callee)) {
+      return true;
+    } else {
+      return false;
```
Comment (edwardlee-msft) [inline packages/react-composites/src/composites/CallComposite/utils/MediaGalleryUtils.ts:183]: nit; something like this
```suggestion
  return !!callees?.filter((callee) => {
    return isPhoneNumberIdentifier(callee) || isMicrosoftTeamsAppIdentifier(callee))
```

**Review by edwardlee-msft — APPROVED** [review]

File: packages/react-composites/src/composites/CallComposite/utils/MediaGalleryUtils.ts [hunk 3]
```diff
@@ -160,3 +166,15 @@ export const createAnnouncementString = (
...
 };
+
+/* @conditional-compile-remove(dtmf-dialer) */
+/**
+ * determines if the media gallery should be replaced by the dtmf dialer
+ * @param callees Target callees to determine if the dtmf dialer should be shown
+ * @returns whether the dialer should be the gallery content or not
+ */
+export const showDtmfDialer = (callees?: CommunicationIdentifier[]): boolean => {
+  return !!callees?.filter((callee) => {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/utils/MediaGalleryUtils.ts:177]: Should you be checking against length here also?
`!![]` will return true

File: packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx [hunk 4]
```diff
@@ -89,9 +95,28 @@ export const CallPage = (props: CallPageProps): JSX.Element => {
...
+  const renderDtmfDialerFromStart = showDtmfDialer(callees);
+
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const [dtmfDialerPresent, setDtmfDialerPresent] = useState<boolean>(renderDtmfDialerFromStart);
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const [initialDialerDismissed, setInitialDialerDismissed] = useState<boolean>(false);
 
   /* @conditional-compile-remove(dtmf-dialer) */
-  const [dtmfDialerPresent, setDtmfDialerPresent] = useState<boolean>(false);
+  useEffect(() => {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx:114]: This useeffect is not obvious on its purpose, add comment around this being used to dismiss this dialer if a second participant joins

File: packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx [hunk 5]
```diff
@@ -89,9 +95,28 @@ export const CallPage = (props: CallPageProps): JSX.Element => {
...
+  const adapter = useAdapter();
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const callees = adapter.getState().targetCallees;
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const renderDtmfDialerFromStart = showDtmfDialer(callees);
+
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const [dtmfDialerPresent, setDtmfDialerPresent] = useState<boolean>(renderDtmfDialerFromStart);
+  /* @conditional-compile-remove(dtmf-dialer) */
+  const [initialDialerDismissed, setInitialDialerDismissed] = useState<boolean>(false);
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/pages/CallPage.tsx:111]: consider a useRef for this property instead. We don't use this property to trigger re-renders. It is only used in the `if` to make sure we only auto-dismiss if the dialer was auto-shown.

(untested example):
```
// We show the dialer if we are in a call with a single callee.
// Track auto dismissing the dialer when a second participant joins the call.
const dialerShouldAutoDismiss = useRef<boolean>(renderDtmfDialerFromStart);

useEffect(() => {
  if(dialerShouldAutoDismiss.current && remoteParticipantsConnected > 1) {
    dialerShouldAutoDismiss.current = false;
    setDtmfDialerPresent(false);
  }
}, [remoteParticipantsConnected]);
```

**Review by JamesBurnside — COMMENTED** [review]
LGTM! Just witing for green CI before hitting approve

**Review by JamesBurnside — APPROVED** [review]

**Pull request merged**